File: daily_python/2872_Maximum_Number_of_K_Divisible_Components.py
import os
import sys
from typing import List

class Solution:
    def dfs(self, node: int) -> int:
        s = self.values[node] % self.k
        self.visited[node] = True
        for neighbor in self.graph[node]:
            if not self.visited[neighbor]:
                s = (s + self.dfs(neighbor)) % self.k
        if s % self.k == 0:
            self.ans += 1
        return s
    
    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        self.ans = 0
        self.graph = [[] for _ in range(n)]
        for start, end in edges:
            self.graph[start].append(end)
            self.graph[end].append(start)
        self.values = values
        self.k = k
        self.visited = [False] * n
        self.dfs(0)
        return self.ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            n = int(lines[0].split('=')[1].strip()[:-1])
            edges = [list(map(int, edge.split(','))) for edge in lines[1].split('=')[1].strip()[2:-3].split('},{')]
            values = list(map(int, lines[2].split('=')[1].strip()[1:-2].split(',')))
            k = int(lines[3].split('=')[1].strip()[:-1])
            dataset.append((n, edges, values, k))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for n, edges, values, k in dataset:
        results.append(solution.maxKDivisibleComponents(n, edges, values, k))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2025/2872_Maximum_Number_of_K_Divisible_Components.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int dfs(int node, int parents, vector<vector<int>>& graph, vector<int>& values, int k, int& component){
        int total = values[node] % k;
        for (auto neighbor : graph[node]){
            if (neighbor != parents) total = (total + dfs(neighbor, node, graph, values, k, component)) % k;
        }
        if (total == 0) component++;
        return total;
    }
    int maxKDivisibleComponents(int n, vector<vector<int>>& edges, vector<int>& values, int k) {
        int component = 0;
        vector<vector<int>> graph(n);
        for (vector<int>& edge : edges){
            int a = edge[0];
            int b = edge[1];
            graph[a].push_back(b);
            graph[b].push_back(a);
        }
        dfs(0, -1, graph, values, k, component);
        return component;
    }
};

int main(){
    int n = 5;
    vector<vector<int>> edges = {{0,2},{1,2},{1,3},{2,4}};
    vector<int> values = {1,8,1,4,4};
    int k = 6;
    // int n = 7;
    // vector<vector<int>> edges = {{0,1},{0,2},{1,3},{1,4},{2,5},{2,6}};
    // vector<int> values = {3,0,6,1,5,2,1};
    // int k = 3;
    // int n = 7;
    // vector<vector<int>> edges = {{0,1},{0,2},{1,3},{1,4},{2,5},{2,6}};
    // vector<int> values = {1000000000,1000000000,1000000000,1000000000,1000000000,1000000000,1000000000};
    // int k = 7;
    Solution S;
    cout << S.maxKDivisibleComponents(n, edges, values, k) << endl;
    return 0;
}